09-maze_solver/progress/internal/solver/explore.go

```diff
@@ -0,0 +1,93 @@
+package solver
+
+import (
+	"fmt"
+	"image"
+	"log/slog"
+	"slices"
+	"sync"
+)
+
+func (s *Solver) listenToBranches() {
+	wg := sync.WaitGroup{}
```

Comment:
I would use `errgroup.Group` where you can later also use `SetLimit`. Also it avoids some footguns related to correct `wg.Add` and `wg.Done` usage.

I would recommend trying to run the solver on `2048x2048` completely blank map without treasure and see how many goroutines will be started. The result might be interesting -- although, I'm not quite sure, how interesting.

Reply:
PS: however, using `SetLimit` implies that `pathsToExplore` queue is made unbounded, otherwise there's an easy chance for a deadlock.

Reply:
One of the ideas we had when we started the chapter was to introduce a pool of workers. We had a version with errgroup.Group, but it required a buffered channel to store all messages that weren't read - and the size of the buffer was the size of the image.  But then the amount of stuff we could fit in this chapter before it became gignormous was exceeded and we went back to unlimited worker pool size, with sync.waitGroup rather than errgroup.SetLimit + buffered channel